### PR TITLE
Integrate CacheSystem inside API-Client

### PR DESCRIPTION
**Pull request by fscarponi**
Introduce caching mechanism inside the PackageSearchApiClient

Added several tests to validate caching mechanisms, including cache hits, cache expiration, and non-exhaustiveness of the cache.

Imported cache library and updated project dependencies. This also includes disabling of certain targets in build configuration.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -10,18 +10,21 @@ kotlin {
...
                 api(npm(packageSearchApiModelsVersions.date.fns))
+                api(kotlinxDocumentStore.browser)
             }
         }
         jvmTest {
             dependencies {
                 implementation(packageSearchApiModelsVersions.junit.jupiter.api)
                 implementation(packageSearchApiModelsVersions.junit.jupiter.params)
                 implementation(packageSearchApiModelsVersions.assertk)
+                implementation(kotlinxDocumentStore.mvstore)
```
Comment (lamba92) [inline build.gradle.kts]: add also `kotlinxDocumentStore.mvstore` in `jvmMain`, instead of `jvmTest`.

File: http/client/build.gradle.kts [hunk 2]
```diff
@@ -23,10 +25,26 @@ kotlin {
                 api(packageSearchApiModelsVersions.kotlinx.serialization.json)
             }
         }
+        commonTest{
```
Comment (lamba92) [inline http/client/build.gradle.kts]: indentation here!

File: http/client/build.gradle.kts [hunk 3]
```diff
@@ -23,10 +25,26 @@ kotlin {
...
+            dependencies{
```
Comment (lamba92) [inline http/client/build.gradle.kts]: indentation here!

File: http/client/src/commonMain/kotlin/cache/CacheEntries.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+internal val DEFAULT_EXPIRATION_TIME: Duration = 12.hours
+
+@Serializable
+internal data class CacheEntry<K, V>(
+    val key: K,
+    val value: V,
+    val createdAt: Instant = Clock.System.now()
+) {
+
+    internal fun isExpired(expirationTime: Duration = DEFAULT_EXPIRATION_TIME): Boolean =
```
Comment (lamba92) [inline http/client/src/commonMain/kotlin/cache/CacheEntries.kt]: Remove the default here.

File: http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt [hunk 5]
```diff
@@ -1,51 +1,95 @@
...
+internal typealias ScrollRequestCacheEntry = CacheEntry<SearchPackagesStartScrollRequest, SearchPackagesScrollResponse>
+internal typealias ApiProjectCacheEntry = CacheEntry<SearchProjectRequest, List<ApiProject>>
+internal typealias MiscellaneousCacheEntry = CacheEntry<String, JsonElement>
+
 public class PackageSearchApiClient(
-    public val endpoints: PackageSearchEndpoints,
+    dataStore: DataStore,
     private val httpClient: HttpClient = defaultHttpClient(),
+    public val endpoints: PackageSearchEndpoints,
+    coroutineScope: CoroutineScope = CoroutineScope(httpClient.coroutineContext),
```
Comment (lamba92) [inline http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt]: The Ktor `HttpClient` is a coroutine scope itself. Just change this line to:
```kotlin
coroutineScope: CoroutineScope = httpClient
```

File: http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt [hunk 6]
```diff
@@ -146,56 +190,200 @@ public class PackageSearchApiClient(
...
-            body = GetPackageInfoRequest(ids),
-            requestBuilder = requestBuilder,
-        ).associateBy { it.id }
+        getPackageInfoByIdHashes(
+            idHashes = ids.map { ApiPackage.hashPackageId(it) }.toSet(),
+            useHashes = false,
+            requestBuilder = requestBuilder
+        )
 
     public suspend fun getPackageInfoByIdHashes(
```
Comment (lamba92) [inline http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:239]: This function should be private. Another one without the parameter `useHashes` should be made publicly available:
```kotlin
public suspend fun getPackageInfoByIdHashes(
    idHashes: Set<String>,
    requestBuilder: (HttpRequestBuilder.() -> Unit)? = null,
): Map<String, ApiPackage> = 
    getPackageInfoByIdHashes(
        idHashes = idHashes,
        useHashes = true,
        requestBuilder = requestBuilder
    )
```

File: http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt [hunk 7]
```diff
@@ -146,56 +190,200 @@ public class PackageSearchApiClient(
...
+        idHashes: Set<String>,
+        useHashes: Boolean = true,
         requestBuilder: (HttpRequestBuilder.() -> Unit)? = null,
-    ): Map<String, ApiPackage> =
-        defaultRequest<_, List<ApiPackage>>(
+    ): Map<String, ApiPackage> = coroutineScope {
+
+        val packagesCache = packagesCacheCollection.await()
+
+        val isOffline = !onlineStateFlow.value
```
Comment (lamba92) [inline http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt]: You can inline this one.
- MacOs shortcut: `cmd` + `option` + `N`
- Windows shortcut: `ctrl` + `alt` + `N`

File: http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt [hunk 8]
```diff
@@ -146,56 +190,200 @@ public class PackageSearchApiClient(
...
         requestBuilder: (HttpRequestBuilder.() -> Unit)? = null,
-    ): SearchPackagesScrollResponse =
-        defaultRequest<_, SearchPackagesScrollResponse>(
+    ): SearchPackagesScrollResponse {
+
+        val scrollCache = scrollRequestCacheCollection.await()
+        scrollCache.find(selector = ScrollRequestCacheEntry::key.name, value = request)
+            .firstOrNull()
+            ?.also {
+                if (!it.isExpired()) return it.value
```
Comment (lamba92) [inline http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt]: One line it! 
```kotlin
?.also { if (!it.isExpired()) return it.value }
```

File: http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt [hunk 9]
```diff
@@ -211,40 +399,54 @@ public class PackageSearchApiClient(
     public suspend fun searchProjects(
         request: SearchProjectRequest,
         requestBuilder: (HttpRequestBuilder.() -> Unit)?,
-    ): List<ApiProject> =
-        defaultRequest<_, List<ApiProject>>(
+    ): List<ApiProject> {
+
+        val apiProjectsCache = apiProjectCacheCollection.await()
+
+        apiProjectsCache.find(selector = SearchCacheEntry::key.name, value = request).firstOrNull()
```
Comment (lamba92) [inline http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt]:
```kotlin
.find(selector = SearchCacheEntry::key.name, value = request)
   .firstOrNull()
   ?.also { if (!it.isExpired()) return it.value }
```
Please!

File: http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt [hunk 10]
```diff
@@ -211,40 +399,54 @@ public class PackageSearchApiClient(
...
             requestBuilder = requestBuilder,
-        )
+        ).also {
+            apiProjectsCache.updateWhere(
+                fieldSelector = ApiProjectCacheEntry::key.name,
+                fieldValue = request,
+                upsert = true,
+                update = ApiProjectCacheEntry(request, it)
+            )
+        }
```
Comment (lamba92) [inline http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:421]: Save also the packages here, just like in `startScroll`.

File: http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt [hunk 11]
```diff
@@ -211,40 +399,54 @@ public class PackageSearchApiClient(
...
         requestBuilder: (HttpRequestBuilder.() -> Unit)? = null,
-    ): List<ApiPackage> =
-        defaultRequest<_, List<ApiPackage>>(
-            method = HttpMethod.Post,
-            url = endpoints.refreshPackagesInfo,
-            body = request,
-            requestBuilder = requestBuilder,
-        )
+    ): List<ApiPackage> {
+        //no caches for this endpoint
```
Comment (lamba92) [inline http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt]: Please put the caches for this as well. Even tho this endpoint does not impact the outbound traffic costs, it still has to be process. Restarting the IDE, re-opening a project, changing the layout of a project and many other actions will still re-trigger this call.

**Review by lamba92 — CHANGES_REQUESTED** [review]

File: http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt (hunk 10 above)
Comment (fscarponi) [reply http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:421]: ApiProject contains only the ids of packages, so we can't

Comment (lamba92) [reply http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:421]: Ah yes, my bad.

**Pull request merged**